### PR TITLE
college app init

### PR DESCRIPTION
**Pull request by druval-cr-zopsmart**
modularize, enable eslint

### CODE REVIEW

**Review by anoophm — APPROVED** [review]

File: druval_cr/CollegeApplication/src/helper/validation.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+  return true;
+}
+
+export function validSubjectRequestData(data: SubjectRequestI) {
+  if (!data) return false;
+  if (data.name && (typeof data.name !== 'string' || !data.name.trim())) return false;
+  return true;
+}
+
+export function validClassRequestData(data: SubjectRequestI) {
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/helper/validation.ts:61]: we can use https://joi.dev/api/?v=17.4.0 for validation, try this out in few api's

File: druval_cr/CollegeApplication/src/controller/class.ts [hunk 2]
```diff
@@ -0,0 +1,135 @@
...
+export async function addClass(ctx: Context) {
+  const requestData: ClassRequestI = ctx.request.body;
+  const validRequestData: boolean = validClassRequestData(requestData) && requestData.name;
+  if (!validRequestData) {
+    ctx.status = 400;
+    ctx.body = 'invalid data';
+    return;
+  }
+  try {
+    await dbSetPath();
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/controller/class.ts]: can you check with Himanshu if he's facing this or how he fixed this

File: druval_cr/CollegeApplication/src/controller/class.ts [hunk 3]
```diff
@@ -0,0 +1,114 @@
...
+    }
+
+    const validClass: boolean = await checkExistByUniqueKeys('class', ['id'], [class_id]);
+    if (!validClass) {
+      ctx.status = 400;
+      ctx.body = 'invalid class id';
+      return;
+    }
+    const paramsData: { page: number; size: number } = validPaginationParams(ctx.request.query);
+    const totalEntry: number = await classService.getClassStudents(class_id, { action: 'count' });
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/controller/class.ts:90]: is there a way to combine this two getClassStudents query. check any examples for postgre pagination

File: druval_cr/CollegeApplication/src/controller/general.ts [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+    const student_id: string = requestData.student_id;
+    const class_name: string = requestData.class_name.toLowerCase();
+
+    if (!uuidValidate(student_id)) {
+      ctx.status = 400;
+      ctx.body = 'invalid uuid';
+      return;
+    }
+
+    const validStudent: boolean = await checkExistByUniqueKeys('student', ['id'], [student_id]);
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/controller/general.ts:43]: if we are using uuid for id, then we dont check for uniqueness

File: druval_cr/CollegeApplication/src/controller/general.ts [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    const validClass: boolean = await checkExistByUniqueKeys('class', ['name'], [class_name]);
+    if (!validClass) {
+      ctx.status = 400;
+      ctx.body = 'invalid class name';
+      return;
+    }
+    const class_id: string = await classService.getClassId(class_name);
+
+    const duplicate: boolean = await checkExistByUniqueKeys('student_class', ['student_id'], [student_id]);
+    if (duplicate) {
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/controller/general.ts:58]: we can keep throwing errors from services with some code and message and use catch to collect and send the error back to client. Check how we can do this.

File: druval_cr/CollegeApplication/src/controller/general.ts [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+  } catch (e) {
+    ctx.status = 404;
+    ctx.body = {
+      message: 'server error',
+      error: e,
+    };
+  }
+}
+
+async function reductiveLeaderboard(allClasses: ClassObjI[], result: any[], leaderboardLength: number) {
```
Comment (anoophm) [inline druval_cr/CollegeApplication/src/controller/general.ts:119]: this can go to service/helper file

**Review by anoophm — CHANGES_REQUESTED** [review]

File: druval_cr/CollegeApplication/src/controller/general.ts (hunk 4 above)
Comment (druval-cr-zopsmart) [reply druval_cr/CollegeApplication/src/controller/general.ts:43]: this is for adding a particular student to a class, so here I was checking whether the student_id sent by the user is a valid id or not.

File: druval_cr/CollegeApplication/src/controller/class.ts (hunk 2 above)
Comment (druval-cr-zopsmart) [reply druval_cr/CollegeApplication/src/controller/class.ts]: this issue is solved in the new PR

**Pull request closed without merging**